onnxruntime/core/providers/webnn/builders/impl/gemm_op_builder.cc

```diff
@@ -268,11 +268,64 @@ bool GemmOpBuilder::HasSupportedInputsImpl(const GraphViewer&, const Node& node,
     return false;
   }
 
-  if (op_type == "MatMulInteger") {
-    // The first decomposed op of MatMulInteger is DequantizeLinear, and so
-    // we only need to ensure it supports the input0_type.
+  if (op_type == "Gemm") {
+    return IsInputRankSupportedByOp(node, wnn_limits, logger) &&
+           IsDataTypeSupportedByOp(op_type, input0_type, wnn_limits, "a", "A", logger);
+  } else if (op_type == "MatMulInteger") {
+    // Check up to 4 inputs for MatMulInteger
+    for (size_t i = 0; i < std::min<size_t>(4, input_defs.size()); ++i) {
+      std::vector<int64_t> shape;
+      if (!GetShape(*input_defs[i], shape, logger)) {
+        LOGS(logger, VERBOSE) << "Failed to get shape for input index " << i;
+        return false;
+      }
+
+      // Inputs 0 and 1 (A and B) support 1D, skip further checks for them if 1D
+      if (i <= 1 && shape.size() == 1) {
+        continue;
+      }
+
+      // For DequantizeLinear, input indices: 0 (x), 1 (scale), 2 (zero_point)
+      int dequant_input_idx = (i < 2) ? 0 : 2;
+      const std::string_view webnn_op_type = "dequantizeLinear";
+      const std::string_view input_name = getWebNNInputName("DequantizeLinear", dequant_input_idx);
+      if (input_name.empty()) {
+        LOGS(logger, VERBOSE) << "Failed to get WebNN input name for DequantizeLinear, input index " << dequant_input_idx << ".";
+        return false;
+      }
+
+      if (!IsInputRankSupported(wnn_limits, webnn_op_type,
+                                input_name,
+                                shape.size(), node.Name(), logger)) {
+        return false;
+      }
+    }
     return IsDataTypeSupportedByOp("DequantizeLinear", input0_type, wnn_limits, "input", "x", logger);
-  } else {
+  } else {  // MatMul
+    for (int i = 0; i < 2; ++i) {
```

Comment:
There're only 2 inputs for MatMul, I don't think we need a for loop.

Reply:
But we might write two duplicate code for `0` & `1`